### PR TITLE
Improve performance of dashboard list orgs

### LINKED ISSUE #15412: Open Pull-Requests view with > 500 organizations need more than 1 minute
- Gitea version (or commit ref): 1.14.0
- Git version:
- Operating system: 20.04.2
  
  
  
- Database (use `[x]`):
  - [x] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
- Log gist:

## Description

This with 1 minute is on my internal mini-server configuration.

I try to analyze what are you doing after I open the Pull-Requests link.

image

What I see.

First this PostgreSQL query:

`explain analyze SELECT "repository"."id" FROM "repository" INNER JOIN "team_user" ON repository.owner_id = team_user.org_id INNER JOIN "team_repo" ON ('true' != 'false' and repository.is_private != 't') OR (team_user.team_id = team_repo.team_id AND repository.id = team_repo.repo_id) WHERE (team_user.uid = '1') GROUP BY repository.id`

image

Which needs ~7 seconds on my NUC (https://ark.intel.com/content/www/us/en/ark/products/126137/intel-nuc-kit-nuc7pjyh.html -> I'm running here Gitea 1.14.0, PostgreSQL 13, Nginx, InfluxDB, Grafana), And after the first query I see a lot small PostgreSQL queries. I think one for every organization.

And when I choose other organization from the long list

image

I see > 500 http requests to the Gitea server

image

Looks like the process is "complicated" and not scalable for a lot of organizations and repos. Cause I have no pull requests on my Gitea instance

image

I think you should return back only all organizations which has pull requests. In my case: no one. Just the info: you dind't have any pull requests.

But this "problem" has only user with a lot of organizations and repos.

Am I right?

### PR DESCRIPTION
**Pull request by lunny**
partially fix #15412

### CODE REVIEW

File: models/org.go [hunk 1]
```diff
@@ -425,6 +425,25 @@ func GetOrgsByUserID(userID int64, showAll bool) ([]*User, error) {
...
 
+// QueryUserOrgIDs returns a condition to return user's organization id
+func QueryUserOrgIDs(uid int64) *builder.Builder {
+	return builder.Select("team.org_id").
+		From("team_user").InnerJoin("team", "team.id = team_user.team_id").
+		Where(builder.Eq{"team_user.uid": uid})
+}
+
+// SimpleOrg represents a simple orgnization with only needed columns
+type SimpleOrg = User
```
Comment (6543) [inline models/org.go]: why not just use User directly?
for later refactors?!?

Comment (lunny) [reply models/org.go]: To notice that the struct only have part fields filled. If we don't add this, we cannot differ the two Users.

Comment (6543) [reply models/org.go]: doesnt it make more sence then to create a new struct type?

Comment (lunny) [reply models/org.go]: I have tried that, but we have to copy some functions from `User`.

Comment (zeripath) [reply models/org.go]: This is why most Java design moved to use POJOs for data with functionality moved to Services eg UserService to do actions on that data. 

Our data structs are too powerful & ultimately too tightly bound to the db interface. `modules/git` is another example and is why creating a second backend was so so hard.

For the moment don't call this "Simple" call it `MinimalUser` or something like that.

**Review by zeripath — APPROVED** [review]

File: routers/web/user/home.go [hunk 2]
```diff
@@ -49,11 +49,12 @@ func getDashboardContextUser(ctx *context.Context) *models.User {
 	}
 	ctx.Data["ContextUser"] = ctxUser
 
-	if err := ctx.User.GetOrganizations(&models.SearchOrganizationsOptions{All: true}); err != nil {
+	orgs, err := models.GetUserOrgsList(ctx.User.ID)
+	if err != nil {
 		ctx.ServerError("GetOrganizations", err)
```
Comment (KN4CK3R) [inline routers/web/user/home.go]: `ctx.ServerError("GetUserOrgsList", err)`

Comment (lunny) [reply routers/web/user/home.go]: done.

File: models/org.go [hunk 3]
```diff
@@ -425,6 +425,25 @@ func GetOrgsByUserID(userID int64, showAll bool) ([]*User, error) {
...
 
+// queryUserOrgIDs returns a condition to return user's organization id
+func queryUserOrgIDs(uid int64) *builder.Builder {
+	return builder.Select("team.org_id").
+		From("team_user").InnerJoin("team", "team.id = team_user.team_id").
+		Where(builder.Eq{"team_user.uid": uid})
+}
+
+// SimpleOrg represents a simple orgnization with only needed columns
+type SimpleOrg = User
```
Comment (6543) [inline models/org.go]:
```suggestion
// MinimalOrg represents a simple orgnization with only needed columns
type MinimalOrg = User
```

File: models/org.go [hunk 4]
```diff
@@ -425,6 +425,25 @@ func GetOrgsByUserID(userID int64, showAll bool) ([]*User, error) {
...
+
+// GetUserOrgsList returns one user's all orgs list
+func GetUserOrgsList(uid int64) ([]*SimpleOrg, error) {
```
Comment (6543) [inline models/org.go]:
```suggestion
func GetUserOrgsList(uid int64) ([]*MinimalOrg, error) {
```

File: models/org.go [hunk 5]
```diff
@@ -425,6 +425,25 @@ func GetOrgsByUserID(userID int64, showAll bool) ([]*User, error) {
...
+		From("team_user").InnerJoin("team", "team.id = team_user.team_id").
+		Where(builder.Eq{"team_user.uid": uid})
+}
+
+// MinimalOrg represents a simple orgnization with only needed columns
+type MinimalOrg = User
+
+// GetUserOrgsList returns one user's all orgs list
+func GetUserOrgsList(uid int64) ([]*MinimalOrg, error) {
+	var orgs = make([]*SimpleOrg, 0, 20)
```
Comment (6543) [inline models/org.go]:
```suggestion
	var orgs = make([]*MinimalOrg, 0, 20)
```

**Review by 6543 — APPROVED** [review]

**Pull request merged**